server/resourcemanager/resource_manager.go

```diff
@@ -235,6 +239,55 @@ func (m *manager[T]) create(w http.ResponseWriter, r *http.Request) {
 	}
 }
 
+func (m *manager[T]) upsert(w http.ResponseWriter, r *http.Request) {
+	encoder := EncoderFromRequest(r)
+
+	targetResource := Resource[T]{}
+	err := encoder.DecodeRequestBody(&targetResource)
+	if err != nil {
+		writeError(w, encoder, http.StatusBadRequest, fmt.Errorf("cannot parse body: %w", err))
+		return
+	}
+
+	if !targetResource.Spec.HasID() {
+		targetResource.Spec = m.rh.SetID(targetResource.Spec, m.config.idgen())
+	}
+
+	writeResponse := func(status int, spec T) {
+		newResource := Resource[T]{
+			Type: m.resourceTypeSingular,
+			Spec: spec,
+		}
+
+		err = encoder.WriteEncodedResponse(w, status, newResource)
+		if err != nil {
+			writeError(w, encoder, http.StatusInternalServerError, fmt.Errorf("cannot marshal entity: %w", err))
+		}
+	}
+
+	_, err = m.rh.Get(r.Context(), targetResource.Spec.GetID())
+	if err != nil {
+		if err == sql.ErrNoRows {
+			created, err := m.rh.Create(r.Context(), targetResource.Spec)
+			if err != nil {
+				writeError(w, encoder, http.StatusInternalServerError, fmt.Errorf("cannot create entity: %w", err))
+				return
+			}
+
+			writeResponse(http.StatusCreated, created)
+			return
+		}
+	}
```

Comment:
Suggestion: we can join the two to avoid nesting:
```suggestion
	if err != nil && err == sql.ErrNoRows {
		created, err := m.rh.Create(r.Context(), targetResource.Spec)
		if err != nil {
			writeError(w, encoder, http.StatusInternalServerError, fmt.Errorf("cannot create entity: %w", err))
			return
		}

		writeResponse(http.StatusCreated, created)
		return
	}
```

Reply:
Your suggestion made me realize I was ignoring the `err` when it was not a `sql.ErrNoRows`